### DISCUSSION TITLE
Passing input ownership?

### DISCUSSION THREAD

**Discussion by albertok** [foxssake/netfox#205] [Q&A] [answered]
I have enterable vehicles in my game that can be entered and driven by any player providing they are empty.

Currently, I've implemented this by having the vehicles as character controllers with RollbackSync and an Input.

 The server owns all empty vehicles and when a player enters one the server gives ownership of the input to the player and they are now in control. When they exit the server takes that control back.

I'm calling `rollback_synchronizer.process_settings()` every time someone enters or exits a vehicle and I've noticed this kind of "shocks" the game a bit. There's a bit of a stutter.

I've seen the docs basically say don't do this and I can see the code basically performs a big wipe on states.

I'm wondering what might be a better way of doing this ownership pass-across?

Would a PR for a smoother ownership handoff method be welcome?

**Comment by elementbound** [comment, marked as answer]
Hi @albertok,

Is the stutter in timing only ( i.e. game freezes for a little bit ), or is it also in state ( i.e. vehicle or player jumps around )? If it's only the former, that could be because of #175, which will be fixed soon by #204. If it's not too much hassle, I'd be happy to hear from you if that version works better for your case.

IIRC the original consideration for not recommending `process_settings()` during gameplay is for cases where people change the properties being synced. That could cause issues that I did not want to support at the time :slightly_smiling_face:  But for authority changes I expect it to be fine. A separate method would be the best, for ownership changes only - that way we can also avoid the state wipe. 

If you'd like to contribute with a PR, that is very welcome! I can also get around to it some time this week.

**Reply by albertok** [reply]
Thanks @elementbound , there's definitely a noticeable difference with the timing fix so that must have played a part.

The final bit to fixing the stutter was to block input for a few ticks to prevent a race condition where clients weren't happy input was coming from someone without ownership yet.

**Reply by elementbound** [reply]
Cool, great to hear! I'll get a release out, probably in a week, with the above fix + an extra method to update ownership only. I'll update this thread too.

**Reply by elementbound** [reply]
Submitted the new version to the asset store today, it'll be available after they approve it. In the meanwhile, the new release is out here on Github.
